### PR TITLE
[ci] prevent Python tests from leaving behind files

### PR DESCRIPTION
**Pull request by vnherdeiro**
Contributes to #6361 

Prevents the following files from being left behind when the Python tests are run:

* `data_dask.csv`
* `Tree4.gv`
* `Tree4.gv.pdf`

The graphivz artefacts from `tests/python_package_test/test_plotting.py` now go to the pytest `tmp_path` and are deleted after test run.

### CODE REVIEW

File: tests/python_package_test/test_plotting.py [hunk 1]
```diff
@@ -257,7 +262,7 @@ def test_tree_with_categories_below_max_category_values():
 
 
 @pytest.mark.skipif(not GRAPHVIZ_INSTALLED, reason="graphviz is not installed")
-def test_tree_with_categories_above_max_category_values():
+def test_tree_with_categories_above_max_category_values(tmp_path: Path):
```
Comment (jameslamb) [inline tests/python_package_test/test_plotting.py:265]:
```suggestion
def test_tree_with_categories_above_max_category_values(tmp_path):
```

Please revert this. We prefer not to use type hints on `pytest` test-case functions in this project.

File: tests/python_package_test/test_plotting.py [hunk 2]
```diff
@@ -213,7 +217,7 @@ def test_create_tree_digraph(breast_cancer_split):
 
 
 @pytest.mark.skipif(not GRAPHVIZ_INSTALLED, reason="graphviz is not installed")
-def test_tree_with_categories_below_max_category_values():
+def test_tree_with_categories_below_max_category_values(tmp_path: Path):
```
Comment (jameslamb) [inline tests/python_package_test/test_plotting.py:220]:
```suggestion
def test_tree_with_categories_below_max_category_values(tmp_path):
```

File: tests/python_package_test/test_plotting.py [hunk 3]
```diff
@@ -1,4 +1,7 @@
 # coding: utf-8
+
+from pathlib import Path
+
```
Comment (jameslamb) [inline tests/python_package_test/test_plotting.py:4]:
```suggestion
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for working on this!

I haven't tested yet if you've caught all the cases where this happens, but these fixes look like what I'd expect. Please see the suggestions I left, and the changes I made to the title and description.

PR titles become commit messages and therefore changelog entries in this repo, so they should describe the changes in a way that would be clear to someone reading the release notes. To see the conventions we follow here, compare these:

* commit log on `master`: https://github.com/microsoft/LightGBM/commits/master/
* most recent release notes: https://github.com/microsoft/LightGBM/releases/tag/v4.5.0

**Comment by vnherdeiro** [comment]
Noted and adressed your comments @jameslamb thanks for the prompt review

**Comment by jameslamb** [comment]
Please update this branch to the latest `master` (to pull in recent CI fixes), and fix the issues reported at https://github.com/microsoft/LightGBM/actions/runs/10605651987/job/29394936451?pr=6626.

You can fix those by running the following from the root of the repo:

```shell
pip install pre-commit
pre-commit run --all-files
```

**Review by borchero — COMMENTED** [review]
Are there any other such occurrences as in #6627 and here? Maybe we can merge all of these changes into a single PR?

**Comment by vnherdeiro** [comment]
@borchero According to https://github.com/microsoft/LightGBM/issues/6361 these are the last two causes of test "leftover" files. I am okay with merging the two PRs.

**Review by borchero — APPROVED** [review]
LGTM :)

**Comment by vnherdeiro** [comment]
Have not merged #6627 as the total code diff is small and they share the same purpose. I let the moderators close the other PR.

**Review by jameslamb — APPROVED** [review]
Ok thanks very much! I've merged in latest `master` to get the changes from #6636, will merge this if CI passes.

We really appreciate the help!

**Comment by vnherdeiro** [comment]
Thanks for the approvals. I am very pleased with this first minor contribution to a package I have been using with enjoyment and satisfaction! Let's see if I can add more to it in the future.

**Pull request merged**